### PR TITLE
Extract generic traits from postgresql structs

### PR DESCRIPTION
**Pull request by tafia**
In order to allow multiple database implementations,
`Connection` and `DbResult` have been promoted to Traits instead of structs

A new folder *postgresql* have been added with the relevant postgresql implementations (PGConnection, PGDbResult and the query builder)

This is a first step toward a generic ORM. Hopefully implementing diesel for new databases will be a matter of adding a new folder + some sql changes.

**NOTE**: I don't have a postgresql installed on my machine so I couldn't run the tests. I even couldn't compile tests because of linking issues. Still I think this is a good first step ....

### CODE REVIEW

**Comment by sgrif** [comment]
Thanks for the PR. There's some failing tests, but I think I can handle getting them passing. I think it's still a bit premature to try and genericise these interfaces, but this all seems pretty reasonable at first glance. This is a pretty hefty PR, so it'll take me a day or two to get through it.

**Comment by tafia** [comment]
Of course I understand.
I didn't really try reasoning about what should be generic or not. My main
focus was to ensure postgresql was in its own module without breaking too
much things.
Take your time.
Thanks

**Comment by tafia** [comment]
Pushed one more commit to relocate postgresql data types and extensions to postgresql directory.
Contrary to `Connection`/`DbResult`, the code was already generic. The only changes are `use` expressions within postgresql folder and lib.rs

**Comment by sgrif** [comment]
Hey just wanted to update why I haven't given this much attention lately. I've started to give more thought to sqlite 3 support, and there's a couple of hard problems that need to be solved.

- Our `NativeSqlType` API only makes sense for PG, but there's not an obvious other place to put it.
- The way we handle bind parameters during query construction works fine for PG and MySQL's API, but is fundamentally opposed to SQLite's.

Both of these lead me to believe that we very well might end up simply having the traits themselves have different APIs based on a `cfg` flag. If that is the route that we end up going down, I'm not sure that it actually makes sense to promote some of these to traits (though it might be worth it any way)

In any case, I'm hesitant to do some of the easier changes until we figure out the harder ones.

**Comment by sgrif** [comment]
Note: I wrote up some very long thoughts on this issue overall in https://github.com/sgrif/diesel/issues/39#issuecomment-172110839

File: diesel/src/connection/cursor.rs [hunk 1]
```diff
@@ -6,15 +6,17 @@ use std::marker::PhantomData;
...
-pub struct Cursor<ST, T> {
+pub struct Cursor<ST, T, R> {
     current_row: usize,
-    db_result: DbResult,
+    db_result: R,
     _marker: PhantomData<(ST, T)>,
 }
 
-impl<ST, T> Cursor<ST, T> {
+impl<ST, T, R> Cursor<ST, T, R> where
```
Comment (sgrif) [inline diesel/src/connection/cursor.rs]: While we may go this route in the future, to limit the scope of this PR, I'd prefer that we did not change this type, and instead had the various methods on `Connection` return `Box<Iterator>`

File: diesel/src/connection/mod.rs [hunk 2]
```diff
@@ -37,27 +21,17 @@ pub type PkType<T> = <PrimaryKey<T> as Expression>::SqlType;
 #[doc(hidden)]
 pub type FindPredicate<T, PK> = Eq<PrimaryKey<T>, <PK as AsExpression<PkType<T>>>::Expression>;
 
-impl Connection {
+pub trait Connection: Send + Drop + Sized {
```
Comment (sgrif) [inline diesel/src/connection/mod.rs]: I don't think we should require `Drop` or `Sized`. I'm torn on requiring `Send`

File: diesel/src/connection/mod.rs [hunk 3]
```diff
@@ -37,27 +21,17 @@ pub type PkType<T> = <PrimaryKey<T> as Expression>::SqlType;
...
+
+    type DbResult: DbResult;
+
+    #[doc(hidden)]
+    fn last_error_message(&self) -> String;
```
Comment (sgrif) [inline diesel/src/connection/mod.rs]: This shouldn't be in the connection interface

File: diesel/src/connection/mod.rs [hunk 4]
```diff
@@ -37,27 +21,17 @@ pub type PkType<T> = <PrimaryKey<T> as Expression>::SqlType;
...
-                    transaction_depth: Cell::new(0),
-                })
-            },
-            _ => {
-                let message = last_error_message(connection_ptr);
-                Err(ConnectionError::BadConnection(message))
-            }
-        }
-    }
+    fn establish(database_url: &str) -> ConnectionResult<Self>;
```
Comment (sgrif) [inline diesel/src/connection/mod.rs]: This shouldn't be in the connection interface.

**Comment by sgrif** [comment]
@tafia Thanks for starting the work on this, but there's a lot of feedback that still needs to be addressed, and we're now at the point where this change is actually needed. I'm going to close in favor of #128.

**Comment by tafia** [comment]
No pb!

**Pull request closed without merging**